internal/cleanups/composite.go

```diff
@@ -1,40 +1,74 @@
 package cleanups
 
-import "errors"
+import (
+	"context"
+	"errors"
+	"time"
+)
 
 type Composite struct {
-	cleanups []func() error
+	cleanups []func(ctx context.Context) error
 }
 
 // Add adds a cleanup to be called.
-func (c *Composite) Add(f func() error) {
+func (c *Composite) Add(f func(ctx context.Context) error) {
 	c.cleanups = append(c.cleanups, f)
 }
 
 // Call calls all cleanups in reverse order and returns an error combining all
 // non-nil errors.
-func (c *Composite) Call() error {
-	err := call(c.cleanups)
+// Context is passed to the cleanups, but its cancellation is disallowed and only
+// values passed via WithValue are retained.
+func (c *Composite) Call(ctx context.Context) error {
+	err := call(ctx, c.cleanups)
 	c.cleanups = nil
 	return err
 }
 
 // Release removes all cleanups, turning Call into a no-op.
 // Caller still can call the cleanups by calling the returned function
 // which is equivalent to calling the Call before Release was called.
-func (c *Composite) Release() func() error {
+func (c *Composite) Release() func(context.Context) error {
 	cleanups := c.cleanups
 	c.cleanups = nil
-	return func() error {
-		return call(cleanups)
+	return func(ctx context.Context) error {
+		return call(ctx, cleanups)
 	}
 }
 
-func call(cleanups []func() error) error {
+func call(ctx context.Context, cleanups []func(ctx context.Context) error) error {
 	var errs error
+	// Disallow cancellation, but pass all values.
+	ctx = notCancellable{parent: ctx}
```

Comment:
Eh, I don't like this, and would prefer to just not pass a context here, but I don't think we want a context to be cancelled here, and we only need the values of the context for the `log.G(ctx)` to work as expected.

Reply:
Hmm, looks like it's in 1.21: https://pkg.go.dev/context@master#WithoutCancel